ts_src/taproot.ts

```diff
@@ -0,0 +1,205 @@
+import assert = require('assert');
+import FastPriorityQueue = require('fastpriorityqueue');
+import * as bcrypto from './crypto';
+const ecc = require('tiny-secp256k1');
+
+// const SECP256K1_ORDER = Buffer.from('fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141', 'hex')
+const INITIAL_TAPSCRIPT_VERSION = Buffer.from('c0', 'hex');
+
+const TAPLEAF_TAGGED_HASH = bcrypto.sha256(Buffer.from('TapLeaf'));
+const TAPBRANCH_TAGGED_HASH = bcrypto.sha256(Buffer.from('TapBranch'));
+const TAPTWEAK_TAGGED_HASH = bcrypto.sha256(Buffer.from('TapTweak'));
+
+interface WeightedTapScript {
+  /** A TapLeaf or TapBranch tagged hash */
+  taggedHash: Buffer;
+  weight: number;
+}
+
+/**
+ * Trims the leading 02/03 byte from an ECDSA pub key to get a 32 byte schnorr
+ * pub key with x-only coordinates.
+ * @param pubkey A 33 byte pubkey representing an EC point
+ * @returns a 32 byte x-only coordinate
+ */
+export function trimFirstByte(pubkey: Buffer): Buffer {
+  assert(pubkey.length === 33);
+  return pubkey.slice(1, 33);
+}
+
+/**
+ * Aggregates a list of public keys into a single MuSig public key
+ * according to the MuSig paper.
+ * @param pubkeys The list of pub keys to aggregate
+ * @returns a 32 byte Buffer representing the aggregate key
+ */
+export function aggregateMuSigPubkeys(pubkeys: Buffer[]): Buffer {
+  // sort keys in ascending order
+  pubkeys.sort();
+
+  const trimmedPubkeys: Buffer[] = [];
+  pubkeys.forEach(pubkey => {
+    const trimmedPubkey = trimFirstByte(pubkey);
+    trimmedPubkeys.push(trimmedPubkey);
+  });
+
+  // In MuSig all signers contribute key material to a single signing key,
+  // using the equation
+  //
+  //     P = sum_i µ_i * P_i
+  //
+  // where `P_i` is the public key of the `i`th signer and `µ_i` is a so-called
+  // _MuSig coefficient_ computed according to the following equation
+  //
+  // L = H(P_1 || P_2 || ... || P_n)
+  // µ_i = H(L || i)
+
+  const L = bcrypto.sha256(Buffer.concat(trimmedPubkeys));
+
+  let aggregatePubkey: Buffer;
+  pubkeys.forEach(pubkey => {
+    const trimmedPubkey = trimFirstByte(pubkey);
+    const c = bcrypto.sha256(Buffer.concat([L, trimmedPubkey]));
+
+    const tweakedPubkey = ecc.pointMultiply(pubkey, c);
+    if (aggregatePubkey === undefined) {
+      aggregatePubkey = tweakedPubkey;
+    } else {
+      aggregatePubkey = ecc.pointAdd(aggregatePubkey, tweakedPubkey);
+    }
+  });
```

Comment:
```suggestion
const tweakedPubkeys = pubkeys .map((pubkey) => {
  const trimmedPubkey = trimFirstByte(pubkey);
  const c = bcrypto.sha256(Buffer.concat([L, trimmedPubkey]));
  return ecc.pointMultiply(pubkey, c);
})
const aggregatePubkey = tweakedPubkeys.slice(1).reduce(ecc.pointAdd, tweakedPubkeys[0]);
```
Alternate construction - kinda two separate ideas:
1) make tweaking and combining the pubkeys separate, because this helps show that order doesn't matter
2) use a slightly clever `reduce` construction to eliminate the ugly conditional

Reply:
I like it. Needed to tweak this slightly though, the suggested code strangely caused the tests to fail. Turns out that `pointAdd` takes a third parameter indicating whether the key is compressed or not, if falsey it returns a 65 byte uncompressed key instead of the 33 byte compressed key. The third parameter for `reduce` is the current index so for index `0`, `pointAdd` treats that as false and returns a 65 byte key and messes the whole thing up. Got it to work with `reduce((prev, curr) => ecc.pointAdd(prev, curr), tweakedPubkeys[0]);`. Had me questioning my sanity briefly though.